
# rGFA graph  
## Pangenome graph not exists  

1. Prepare assemblies 

```
sh download.testData.sh

sh prepare.asm.sh
```
2. Construct a pangenome graph and build index  

```
python3 ../script/vrpg_preprocess.py --minigraph ../bin/minigraph --asmList build.asm.txt --index --xDep 50 --outDir results
```
3. Extract and index node sequences  
```
../module/nodeSeq --gfaFile results/upload/input.ref.gfa --upDir  results/upload
```
4. Add primary reference gene track  

```
../module/GraphAnno addRef --inGFF GCF_000146045.2_R64_genomic.gff.gz --upDir results/upload
```

5. Add annotation for all nodes  
```
sh prepare.gff.sh

../module/GraphAnno nodeGene --gffList build.gff.txt --upDir results/upload
```

6. Add annotation track from BED file  
```
../module/GraphAnno addBed --inBed test.track.bed --upDir results/upload
```

7. Migrate data  
```
mv results/upload/* ../upload
```

8. Start the development server  

```
python3 manage.py runserver
```
9. Visit pangenome regions  

Open http://127.0.0.1:8000/app/vrpg/ in your browser.    

## Pangenome graph alreaddy exists  

We will use the files generated by running steps given above to show how to use VRPG on the condition that graph was already constructed.  

1. Align assemblies to the graph

```
# Let's assume that data migration has not been done and input.ref.gfa still in results/upload.
mkdir mapping
minigraph -cxasm -t 10 --vc results/input.ref.gfa DBVPG6044.genome.fa.gz -o mapping/DBVPG6044.unstable.gaf
minigraph -cxasm -t 10 --vc results/input.ref.gfa DBVPG6765.genome.fa.gz -o mapping/DBVPG6765.unstable.gaf
minigraph -cxasm -t 10 --vc results/input.ref.gfa SK1.genome.fa.gz -o mapping/SK1.unstable.gaf
minigraph -cxasm -t 10 --vc results/input.ref.gfa Y12.genome.fa.gz -o mapping/Y12.unstable.gaf

```
2. Create GAF file list  

```
echo "DBVPG6765#HP0 mapping/DBVPG6765.unstable.gaf NA" >> mapping/gaf.list
echo "Y12#HP0 mapping/Y12.unstable.gaf NA" >> mapping/gaf.list
echo "SK1#HP0 mapping/SK1.unstable.gaf NA" >> mapping/gaf.list
echo "DBVPG6044#HP0 mapping/DBVPG6044.unstable.gaf NA" >> mapping/gaf.list
```

3. Build index

```
python ../script/vrpg_preprocess.py --gafList mapping/gaf.list --rGFA results/input.ref.gfa --outDir rGFA_upload --thread 5 --index --xDep 100
```

Remaining steps are same as that graph has not been constructed.  
 
# GFA graph  
## Minitraph-Cactus graph  
1.	Construct a pangenome graph by using Minitraph-Cactus  

The parameters and options used to construct the test graph may not be suitable for all cases. About how to construct a pangenome graph using Minigraph-Cacuts, please read the tutorial (https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/pangenome.md) of Minigraph-Cacuts.

```
# Assume that Minitraph-Cactus has been installed
sh mc.genome.sh
cactus-pangenome ./js mc.genome.txt --outDir mc --outName mc --reference SGDref --gfa

```

2. Transform and index the graph  

```
../module/gfa2view --GFA mc/mc.gfa.gz --ref SGDref#0 --outDir mc_upload --index --range 2000 --thread 5 --xDep 100

```

3. Extract sequence and add annotation tracks

```
sh mc.gff.sh

../module/nodeSeq --gfaFile mc/mc.gfa.gz --upDir  mc_upload/upload
../module/GraphAnno addRef --inGFF GCF_000146045.2_R64_genomic.gff.gz --upDir mc_upload/upload
../module/GraphAnno nodeGene --gffList mc.gff.txt --upDir mc_upload/upload
../module/GraphAnno addBed --inBed test.track.bed --upDir mc_upload/upload

```

4. Migrate data

```
mv mc_upload/upload ../upload/mc_graph
```

Other steps are similar to that for rGFA graph.  

## PGGB graph  

1. Construct a pangenome graph by using PGGB  

The parameters and options used to construct the test graph may not be suitable for all cases. About how to construct a pangenome graph using PGGB, please read the tutorial (https://github.com/pangenome/pggb) of PGGB.

```
# Assume that fastix and pggb have been installed
sh pggb.genome.sh
pggb -i pggb_genome/all.fastix.fa -t 10 -p 95 -n 5 -k 23 -o pggb

```
2. Transform and index the graph  

```
# Please replace the * in 'all.fastix.fa.*.final.gfa' with the actual file name
../module/gfa2view --GFA pggb/all.fastix.fa.*.final.gfa --ref SGDref#1 --outDir pggb_upload --index --range 2000 --thread 5 --xDep 100

```

3. Extract sequence and add annotation tracks

```
sh pggb.gff.sh

# Please replace the * in 'all.fastix.fa.*.final.gfa' with the actual file name
../module/nodeSeq --gfaFile pggb/all.fastix.fa.*.final.gfa --upDir pggb_upload/upload
../module/GraphAnno addRef --inGFF GCF_000146045.2_R64_genomic.gff.gz --upDir pggb_upload/upload
../module/GraphAnno nodeGene --gffList pggb.gff.txt --upDir pggb_upload/upload
../module/GraphAnno addBed --inBed test.track.bed --upDir pggb_upload/upload
```

4. Migrate data  

```  
mv pggb_upload/upload ../upload/pggb_graph
```

Other steps are similar to that for rGFA graph.  






