# VRPG
an interactive web viewer for reference pangenome graphs

|  |  |
| --- | --- |
| <img src="https://github.com/codeatcg/VRPG/blob/main/static/images/window2.png" width=600px/> | <img src="https://github.com/codeatcg/VRPG/blob/main/static/images/description.png" width=400px/> |

Figrue on the left was based on vrpg-v0.1.1 and the particular assembly path was not highlighted by arrow lines. Figrues on the right were picked from drawings by vrpg-v0.1.2.

# Description  
VRPG is an interactive web viewer for pangenome graph. It was initially designed to navigate the whole pangenome graph in reference Graph Fragment Assembly (<a href="https://github.com/lh3/gfatools/blob/master/doc/rGFA.md">rGFA</a>) format. Now Graph Fragment Assembly (<a href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md">GFA v1.x</a>) format is also supported. This is implemented by converting the GFA format to a format similar to rGFA. VRPG is fast, even for large and complex human pangenome graph generated with many whole genome assemblies. The reference nodes are aligned along the center line of the view window and surrounded by non-reference nodes. The reference line that consists of a series of reference nodes holds the stable coordinate system, just like the traditional linear reference. It’s beneficial to scanning all kinds of variations in the graph and relating them to biological function from previous studies based on a single linear reference. A website shipping four pangenome graphs (one for yeast and three for human) is available at https://www.evomicslab.org/app/vrpg/. The Saccharomyces cerevisiae pangenome graph was generated with 163 assemblies and The three Homo sapiens pangenome graphs with 90 whole genome assemblies for each were constructed by <a href="https://humanpangenome.org/">HPRC</a>  by three different pipelines (Minigraph, Minigraph-CACTUS and PGGB). Users can also deploy the web application and view their own data.  

**Note:** the released version 0.1.2 is not the latest. For graph in GFA format the overlap field in link line (overlap between segments)  should be specified (in graphs created by Minigraph-CACTUS and PGGB the overlap is generally specified as 0M), or the value will be set to 0 by VRPG-gfa2view. Although whether the overlap is specified doesn't affect the visualization of the graph it may affect the determination of the coordinate of the segment. 

The graphs created by Minigraph-CACTUS and PGGB include large amounts of SNPs and INDELs. The structure variations may be covered up by these small variants. The latest version of VRPG supplied functions to simplify the graph, i.e. remove nodes related to the small variants (with size < 50 bp). The simplification related option 'non-ref' in combobox means simplifying non-reference nodes. 'all node' means simplifying all nodes including reference and non-reference. 'none' means not simplifying the graph. Now VRPG can be used to visualize variants at different scales and find their coordinates relative to the reference  conveniently. Furthermore, the function that serves several pangenomes were added back in the latest version of VRPG.

For cola layout the node size is more proportional to the segment sequence size. But it may take a little longer time to stabilize. When the number of nodes in a window is small cola layout can be tested.  

# Installation  
Python3 (>=3.6) and pip environment are required.  

```
# For installing a historical version please access https://github.com/codeatcg/VRPG/releases and download the source code.

# install the latest version
# gcc >= 4.9
pip install Django==3.2.4  pybind11
git clone https://github.com/codeatcg/VRPG --recursive  
cd VRPG/module
make

```

# Prepare your data  

The naming scheme of assembly should follow <a href="https://github.com/pangenome/PanSN-spec">PanSN prefix naming pattern</a>. Briefly, the assembly's name consists of sample name, delimiter, and haplotype name, e.g., sampleA#0. But it's a little looser in VRPG. It's not required that the haplotype name must be numeric, characters are also allowed.  

## rGFA format graph  

### Pangenome graph already exists  

The assemblies to graph mapping files are required. If these files do not exist the assembly can't be highlighted in the drawing. These files can be generated by minigraph by using command '-cxasm --vc'. Then run the following command to get files required by VRPG.  

```
Python script/vrpg_preprocess.py --rGFA all.gfa --gafList gaf_file.list --outDir out_folder --index
```

#### gaf_file.list file is formatted as follows: 

sample1#H1	sample1.H1.gaf  
sample2#0	sample2.0.gaf  
sample3#1	sample3.1.gaf  
sample3#2	sample3.2.gaf  

### Pangenome graph not exists  

Run the following command to create pangenome graph and generate files required by VRPG.  

```
Python script/vrpg_preprocess.py –-minigraph '/software/minigraph' --assList ass_file.list –-outDir out_folder --index
```

#### ass_file.list file is formatted as follows:  
sample1#H1	sample1.H1.fa  
sample2#0	sample2.0.fa  
sample3#1	sample3.1.fa  
sample3#2	sample3.2.fa  

**Note**, '/software/minigraph' represents the absolute path of minigraph executable file. Assembly in first line in file ass_file.list will be taken as reference.  

## GFA format graph

For graphs in GFA format that can be processed by VRPG segment names should be numeric. Fortunately, graphs generated by Minigraph-CACUTUS and PGGB have this feature. If the segment names are not numeric users need to modify the graph first. Also notice that all path names in the graph should follow <a href="https://github.com/pangenome/PanSN-spec">PanSN prefix naming pattern</a>. If the path names don’t obey the rule the graph needs to be modified. This can be avoided by using proper assembly names before constructing the graph. If the graph satisfied the conditions described above run the following command to get files required by VRPG.  

```
module/gfa2view --GFA in.gfa --ref refName --outDir output_dir --index --range 2000 --thread 10

```
**Note**, For the current version of 'gfa2view' memory consumption is proportional to number of threads. A trade-off between speed and and memory consumption needs to be considered.  

If only a particular set of reference chromosomes or contigs are considered for view the option ‘--refChr’ can be used to save running time. For this option a file containing a reference chromosome or contig name (not contain sample name or haplotype name) per line is required.  


# Execution  
## Local server or personal computer with Linux/Ubuntu operating system  

1. Move all output files in directory 'upload' generated during data preparation to the empty folder 'upload' of VRPG. If more than one pangenomes are available users can rename the results directory 'upload' generated during data preparation and then move the renamed directory into the VRPG 'upload' folder.

2. Start the development server of Django  

```
python3 manage.py runserver  

If all is well you will see the output:  
Django version 3.2.4, using settings 'primers_project.settings'  
Starting development server at http://127.0.0.1:8000/  
```

3. Then open http://127.0.0.1:8000/app/vrpg/ in your browser and you will see the drawing of the pangenome graph.  

**Note**, For large pangenome graph it's better to prepare data in a computing server and then transfer the data to local 'upload' folder of VRPG.

## Remote server  

If server is running on a different machine start the server by running

```
python3 manage.py runserver 0.0.0.0:8000
```
Please make sure the firewall is closed. Then open <a>http:\<IP of server\>:8000/app/vrpg/</a>.

If you are familiar with nginx or apache you can also deploy VRPG by using any of them.

# Tips  
1. The edge can be highlighted by clicking on it. Remove the highlight by just click on it again.
2. For graphs generated by PGGB the duplicate sequence in assigned reference assembly may be collapsed. For the collapsed region new nodes and edges will be inserted into the graph by vrpg to make the reference keeping coordinate system same to the traditional single linear reference. But the new nodes and edges will be not inserted into the path. If there are only reference nodes in a region and the highlighted reference path don’t pass these nodes the region may indicate a collapse. Another feature of these regions is that the numeric identifier of the segment that the node represents is generally much larger than the identifiers of the reference segments in the surrounding region.  
<img src="https://github.com/codeatcg/VRPG/blob/main/static/images/ref_collapse3.png"/>
3. VRPG uses 1-based coordinate system, i.e. coordinate is denoted as [start,end]. If you are interest in a particular segment (node) and the genome annotation file (GFF3 format) is available you can use <a href="https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html">bedtools</a> and the coordinate displayed by VRPG (within Segment Source:) to find the annotation of the segment conveniently. For example:
  
  
```  
bedtools intersect -wa -wb -a node.pos -b genome.gff3
``` 
  


 
